test/denormalized_has_many_test.rb

```diff
@@ -141,6 +141,55 @@ def test_deprecated_set_inverse_association_on_cache_hit
     assert_equal item.object_id, associated_record.item.object_id
   end
 
+  def test_returned_records_should_be_readonly_on_cache_hit
+    IdentityCache.with_fetch_read_only_records do
+      Item.fetch(@record.id) # warm cache
+
+      record_from_cache_hit = Item.fetch(@record.id)
+      record_from_cache_hit.fetch_associated_records.each do |record|
+        assert record.readonly?, "Record #{record} wasn't readonly"
+      end
+    end
+  end
+
+  def test_returned_records_should_be_readonly_on_cache_miss
+    IdentityCache.with_fetch_read_only_records do
+      record_from_cache_miss = Item.fetch(@record.id)
+
+      record_from_cache_miss.fetch_associated_records.each do |record|
+        assert record.readonly?, "Record #{record} wasn't readonly"
+      end
+    end
+  end
+
+  def test_db_returned_records_should_never_be_readonly
+    IdentityCache.with_fetch_read_only_records do
+      record_from_db = Item.find(@record.id)
+
+      uncached_records = record_from_db.associated_records
+      uncached_records.each do |record|
+        assert !record.readonly?, "Record #{record} was readonly"
```

Comment:
This isn't really testing anything identity cache, since `Item.find` or `record_from_db.associated_records` are active record methods.  Perhaps you want Item.fetch instead to test a cache miss